### ISSUE TITLE
[REVIEW]: PySDM v1: particle-based cloud modelling package for warm-rain microphysics and aqueous chemistry

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3219]
**Submitting author:** @piotrbartman (<a href="http://orcid.org/0000-0003-0265-6428">Piotr Bartman</a>)
**Repository:** https://github.com/atmos-cloud-sim-uj/PySDM.git
**Branch with paper.md** (empty if default branch): 
**Version:** v1.27
**Editor:** @dhhagan
**Reviewers:** @darothen, @josephhardinee
**Archive:** 10.5281/zenodo.6321270

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9/status.svg)](https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9"><img src="https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9/status.svg)](https://joss.theoj.org/papers/62cad07440b941f73f57d187df1aa6e9)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/atmos-cloud-sim-uj/PySDM.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@josephhardinee & @darothen, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dhhagan know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/atmos-cloud-sim-uj/PySDM.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@piotrbartman) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/atmos-cloud-sim-uj/PySDM.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@piotrbartman) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jwuttke, @darothen it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3219 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.41 s (534.1 files/s, 31905.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                         212           2053           1184           7548
SVG                              1              0             19           1402
Jupyter Notebook                 3              0            207            336
Markdown                         1             38              0            280
YAML                             4             20              4            112
-------------------------------------------------------------------------------
SUM:                           221           2111           1414           9678
-------------------------------------------------------------------------------


Statistical information for the repository 'ee5fa7f3802655beaf36d792' was
gathered on 2021/04/26.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Anna Jaruga                      1             3              1            0.01
Bartosz Piasecki                 1            42             23            0.08
Michael                         18           443            170            0.80
Michaeldz36                     37          2661           2424            6.64
Piotr Bartman                   10            14             19            0.04
Sylwester Arabas               417         11939           7609           25.54
piotrbartman                    61          2072           2853            6.44
prbartman                      438         26379          19740           60.27
tehAgitto                        4            84             51            0.18

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Bartosz Piasecki             30           71.4          5.1                0.00
Michael                     289           65.2         15.1               10.03
Sylwester Arabas           4438           37.2          3.2                4.35
piotrbartman               6011          290.1          9.3                7.97
tehAgitto                    42           50.0          4.5                2.38
```

**Comment by slayoo** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by darothen** [comment]
# Review - @darothen 

`PySDM` is a new open-source package for particle-resolved modeling of the physics and chemistry of cloud and aerosol microphysics. It provides high-performance reference implementations of state-of-the-science process models in a flexible, highly-configurable package that is relatively easy to manipulate and modify to suit a wide variety of cloud and aerosol modeling tasks. It also ships with a back-end for running certain processes on GPUs, although this functionality does not seem to be fully implemented for all configurations which a user may seek to leverage. 

Overall this is a high quality and valuable contribution to the field. Many researchers leverage simple box and parcel models for critical work into fundamental aerosol/cloud microphysics, and `PySDM` provides extremely useful tooling for exploring this domain and for replicating the results of other researchers. However, the library is lacking in documentation, both formally (in clearly defining the complex structure of the objects which comprise the code and how they interact) and informally (through docstrings and other comments/annotations in the code which may allow users to better understand arguments to key functions and how to manipulate and extend them). At present, the authors provide a nice "Hello world" example with some prose on how to configure it, as well as a basic section on package structure and API. But a more comprehensive listing of the arguments, attributes, and parameters that the user should know about in order to configure or extend this example would go a very long way to promote practical usability. 

I would encourage the authors to improve or augment this documentation before the final publication of this software paper. However, in all other aspects I find this to be a comprehensive and uniquely valuable software contribution to the cloud and aerosol microphysical modeling community and an ideal candidate for publication in JOSS.

Please note the handful of detailed comments below on specific changes necessary to meet JOSS standards.

## General Checks

- Two authors on the paper (Grzegorz Lazarski and Aleksandra Talar) do not seem to appear as [contributors](https://github.com/atmos-cloud-sim-uj/PySDM/graphs/contributors) to the repo and aren't listed in "Author Contributions." I don't really believe there is any issue here - for long-incubated software projects like this with lots of research interfaces there have to be many contributors over time! But I do want to note this for the editors. 

## Functionality

- A simple install recipe using PyPI is provided, but given the capability to run on the GPU a `conda-forge` recipe or something which can help readers understand requirements - or provide a turnkey installation with necessary dependencies - would be very helpful! (Issue [494](https://github.com/atmos-cloud-sim-uj/PySDM/issues/494))

- Installation using the documented instructions misses out on installing `pystrict`, a required dependency (Issue/PR [495](https://github.com/atmos-cloud-sim-uj/PySDM/pull/495))

## Documentation

- The automated tests seem to work out of the box, but the smoke tests seem to take a *very* long time to run. Even one of the unit tests (namely `attributes/chemistry/test_pH.py`) took a disproportionate amount of time compared to the others. Please note that I was not able to run the GPU tests as I did not have immediate access to a GPU-powered machine.

- While there is extensive, excellent demonstration and example code available from the authors in related, clearly-linked repositories, the API documentation leaves much to be desired. There is a high degree of customizability and configuration exposed to the user, but very little in the way of curated documentation for the APIs and data structures in the software, or docstrings and annotations. As a motivating example, consider the [spectra](https://github.com/atmos-cloud-sim-uj/PySDM/blob/master/PySDM/initialisation/spectra.py) module; in this module, a nice base class for building probability distributions is exposed, but there are no docstrings describing the required and optional arguments for either the base class or any of the sub-classes which implement it. An effort to improve the documentation with basic docstrings for all the core methods and functionality would greatly improve the quality of the software and the user experience when adapting it for a specialized research task. Ideally, a simple Sphinx-based documentation for the key API and configurations exposed to users would follow very plainly from such an effort, greatly improving the ease of access of the software.

- Community / contribution guidelines are not provided with the repo; the JOSS reviewer guidelines suggest that such documentattion should clearly indicate how third-parties may contribute to the software, report issues, and seek support should they need to do so.

## Software Paper

- The software paper does not include a section which explicitly explains a "Statement of Need." However, I do believe that the prose in the paper covers this topic with the necessary detail. I will defer to the editors whether a stand-alone "Statement of Need" section is required to meet JOSS guidelines.
  
- It's much appreciated that the authors include links to open source software which attempts to tackle subsets of the broad functionality that `PySDM` provides. I would strongly encourage the author to include a brief discussion of each package (particularly `pyrcel` and `PyBox`) and compare/contrast the unique or improved functionalities of `PySDM` relative to each. My take (as the author of one of those libraries, `pyrcel`!) is that `PySDM` offers a substantial improvement over each in terms of the breadth of calculations and science/research that can be performed with it as well as the cleanliness of its interface and design for future improvements and extensions. These are all very positive things for `PySDM` and I would strongly encourage the authors to lean into the fantastic work here to promote the capabilities and utility of their library relative to the competition!

**Comment by slayoo** [comment]
@darothen, on behalf of all authors, let me thank for the feedback! 

Let me just address one comment right away:

> Two authors on the paper (Grzegorz Lazarski and Aleksandra Talar) do not seem to appear as contributors to the repo and aren't listed in "Author Contributions." I don't really believe there is any issue here - for long-incubated software projects like this with lots of research interfaces there have to be many contributors over time! But I do want to note this for the editors.

GŁ's (@golui) and AT's (@aleksandratal) contributions were subject to refactors during which we've not preserved commit authorship, yet these can be traced back, e.g., here: https://github.com/atmos-cloud-sim-uj/PySDM/pull/157 & https://github.com/atmos-cloud-sim-uj/PySDM/pull/142 Both are mentioned in the "Author Contributions" section:

![image](https://user-images.githubusercontent.com/89685/116309753-71790400-a7a9-11eb-9012-cde6f6b06311.png)
 
Thanks!

**Comment by slayoo** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by slayoo** [comment]
> @whedon generate pdf from branch JOSS

This was just to reflect an API change within one of the code snippets presented in the text:
https://github.com/atmos-cloud-sim-uj/PySDM/commit/ca4d8214ce49abeb6b6b08892428eb04b794bc1f

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:wave: @jwuttke, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @darothen, please update us on how your review is going (this is an automated reminder).

**Comment by darothen** [comment]
@dhhagan Just checking in here - anything immediately needed from me here? 

@slayoo please let me know if there are any PRs or other updates I've missed.

**Comment by slayoo** [comment]
We are almost done with preparing a release and will soon after post here a point-by-point reply to the review detailing how the points were addressed in the release. Thanks, S.

**Comment by dhhagan** [comment]
@darothen Thanks for checking in -> you should be good for now. Just waiting on the other reviewer and updated/patches from the authors before any next steps are needed. Thanks for checking in!

**Comment by dhhagan** [comment]
@jwuttke Can you please update us on the status of your review? Thanks!

**Comment by piotrbartman** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by dhhagan** [comment]
@whedon reming @jwuttke in 1 day

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by dhhagan** [comment]
@whedon remind @jwuttke in 1 day

**Comment by whedon** [comment]
Reminder set for @jwuttke in 1 day

**Comment by whedon** [comment]
:wave: @jwuttke, please update us on how your review is going (this is an automated reminder).

**Comment by dhhagan** [comment]
Hey @piotrbartman sorry about the long delays here. I am going to start looking for a new reviewer so we can finish this up. Hope to be back with some good news later this week or next.

**Comment by piotrbartman** [comment]
@dhhagan Thanks for the update!

**Comment by slayoo** [comment]
Let me provide a point-by-point reply to @darothen's review.

First of all, on behalf of all authors, let me express again our appreciation for the feedback. We have been working on addressing the points raised in the review, and a lot is already reflected in the recent 1.4 release of PySDM (https://github.com/atmos-cloud-sim-uj/PySDM/releases/tag/v1.4).

> Review - @darothen
>
> PySDM is a new open-source package for particle-resolved modeling of the physics and chemistry of cloud and aerosol microphysics. It provides high-performance reference implementations of state-of-the-science process models in a flexible, highly-configurable package that is relatively easy to manipulate and modify to suit a wide variety of cloud and aerosol modeling tasks. It also ships with a back-end for running certain processes on GPUs, although this functionality does not seem to be fully implemented for all configurations which a user may seek to leverage.

As correctly pointed out by the reviewer, at the time of the initial submission, the GPU support has been limited to the coalescence process. As of version 1.4, this is extended to all but aqueous chemistry processes, in particular supporting condensational growth/evaporation of particles (yet without adaptive timestepping which features logic less applicable in the GPU SIMT context). Parcel and single-column examples work OK on the GPU, and testing of the GPU code (through the FakeUnits mocking layer) is part of the Continuous Integration workflow.

> Overall this is a high quality and valuable contribution to the field. Many researchers leverage simple box and parcel models for critical work into fundamental aerosol/cloud microphysics, and PySDM provides extremely useful tooling for exploring this domain and for replicating the results of other researchers. However, the library is lacking in documentation, both formally (in clearly defining the complex structure of the objects which comprise the code and how they interact) and informally (through docstrings and other comments/annotations in the code which may allow users to better understand arguments to key functions and how to manipulate and extend them). At present, the authors provide a nice "Hello world" example with some prose on how to configure it, as well as a basic section on package structure and API. But a more comprehensive listing of the arguments, attributes, and parameters that the user should know about in order to configure or extend this example would go a very long way to promote practical usability.

The steps taken so far to improve the documentation include:
- deployment of [pdoc](https://pdoc3.github.io/pdoc/)-generated API manual to GitHub pages at https://atmos-cloud-sim-uj.github.io/PySDM/;
- addition of a parcel-framework condensation-only example to the README file (with code snippets in Matlab, Julia and Python, as for the earlier available coalescence-only example);
- cleanup in docstrings in numerous files in the project.

> I would encourage the authors to improve or augment this documentation before the final publication of this software paper. However, in all other aspects I find this to be a comprehensive and uniquely valuable software contribution to the cloud and aerosol microphysical modeling community and an ideal candidate for publication in JOSS.

Extending the documentation is going to be a priority for the upcoming releases and the growing user-base will help us to fulfill this commitment.

> Please note the handful of detailed comments below on specific changes necessary to meet JOSS standards.
> General Checks
>
>    Two authors on the paper (Grzegorz Lazarski and Aleksandra Talar) do not seem to appear as contributors to the repo and aren't listed in "Author Contributions." I don't really believe there is any issue here - for long-incubated software projects like this with lots of research interfaces there have to be many contributors over time! But I do want to note this for the editors.

As pointed out above (https://github.com/openjournals/joss-reviews/issues/3219#issuecomment-827915407), contributions of both Grzegorz Łazarski and Aleksandra Talar can be traced back on github, yet the commit authorship was partly lost as a part of refactors when handling their contributions. 

>Functionality
>
>    A simple install recipe using PyPI is provided, but given the capability to run on the GPU a conda-forge recipe or something which can help readers understand requirements - or provide a turnkey installation with necessary dependencies - would be very helpful! (Issue 494)

We are actively working on addressing it:
- following our request, the conda ThrustRTC package has been updated: https://github.com/fynv/ThrustRTC/issues/13
- following our request, a conda CURandRTC package has been created: https://github.com/fynv/CURandRTC/issues/3
- a draft of conda meta file is already being worked on in a PySDM PR by @Golui (Grzegorz Łazarski): https://github.com/atmos-cloud-sim-uj/PySDM/pull/575

>    Installation using the documented instructions misses out on installing pystrict, a required dependency (Issue/PR 495)

As elaborated on in the issue (https://github.com/atmos-cloud-sim-uj/PySDM/pull/495):
- the single use of `pystrict` was removed;
- in order to more clearly distinguish between run-time and test-time requirements (listed in `setup.py` and `requirements.txt` file, respectively), the `requirements.txt` file was renamed to `test-time-requirements.txt`;
- package installation via pip on a fresh system was added to the CI workflow to detect such issues automatically.
Thank you for pointing this out!

> Documentation
>
>    The automated tests seem to work out of the box, but the smoke tests seem to take a very long time to run. Even one of the unit tests (namely attributes/chemistry/test_pH.py) took a disproportionate amount of time compared to the others. Please note that I was not able to run the GPU tests as I did not have immediate access to a GPU-powered machine.

The aqueous chemistry code has received a major refactor and cleanup as a part of developments leading to v1.4. The timing of `test_pH.py` has improved. Part of resource usage in `test_pH.py` is taken by `chempy` - an external package used from within the test to compare PySDM results with an independent code.

>    While there is extensive, excellent demonstration and example code available from the authors in related, clearly-linked repositories, the API documentation leaves much to be desired. There is a high degree of customizability and configuration exposed to the user, but very little in the way of curated documentation for the APIs and data structures in the software, or docstrings and annotations. As a motivating example, consider the spectra module; in this module, a nice base class for building probability distributions is exposed, but there are no docstrings describing the required and optional arguments for either the base class or any of the sub-classes which implement it. An effort to improve the documentation with basic docstrings for all the core methods and functionality would greatly improve the quality of the software and the user experience when adapting it for a specialized research task. Ideally, a simple Sphinx-based documentation for the key API and configurations exposed to users would follow very plainly from such an effort, greatly improving the ease of access of the software.

Besides following the suggestion of using a Sphinx-like tool (pdoc3 in this case) and setting up automatic deployment of up-to-date API docs on Github pages as pointed above, we have also reorganised the physics-related code (which is likely to be the first a user would intend to study and extend) and grouped it in the `physics` directory/subpackage with the following structure:
```
    PySDM.physics.aqueous_chemistry
    PySDM.physics.coalescence_kernels
    PySDM.physics.condensation_coordinate
    PySDM.physics.constants
    PySDM.physics.diffusion_kinetics
    PySDM.physics.diffusion_thermics
    PySDM.physics.dimensional_analysis
    PySDM.physics.drop_growth
    PySDM.physics.formulae
    PySDM.physics.hydrostatics
    PySDM.physics.hygroscopicity
    PySDM.physics.impl
    PySDM.physics.latent_heat
    PySDM.physics.particle_advection
    PySDM.physics.saturation_vapour_pressure
    PySDM.physics.spectra
    PySDM.physics.state_variable_triplet
    PySDM.physics.surface_tension
    PySDM.physics.terminal_velocity
    PySDM.physics.trivia
    PySDM.physics.ventilation
```

All of the above subpackages have at least top-level docstrings included.

>    Community / contribution guidelines are not provided with the repo; the JOSS reviewer guidelines suggest that such documentattion should clearly indicate how third-parties may contribute to the software, report issues, and seek support should they need to do so.

A "Contributing, reporting issues, seeking support" section has been added to the README file: https://github.com/atmos-cloud-sim-uj/PySDM/pull/579

> Software Paper
>
>    The software paper does not include a section which explicitly explains a "Statement of Need." However, I do believe that the prose in the paper covers this topic with the necessary detail. I will defer to the editors whether a stand-alone "Statement of Need" section is required to meet JOSS guidelines.

While the "Statement of Need" section title is not used, the Introduction section was meant to cover it.

>    It's much appreciated that the authors include links to open source software which attempts to tackle subsets of the broad functionality that PySDM provides. I would strongly encourage the author to include a brief discussion of each package (particularly pyrcel and PyBox) and compare/contrast the unique or improved functionalities of PySDM relative to each. My take (as the author of one of those libraries, pyrcel!) is that PySDM offers a substantial improvement over each in terms of the breadth of calculations and science/research that can be performed with it as well as the cleanliness of its interface and design for future improvements and extensions. These are all very positive things for PySDM and I would strongly encourage the authors to lean into the fantastic work here to promote the capabilities and utility of their library relative to the competition!

While wholeheartedly agreeing that a comparison of features of different parcel models available in the Python ecosystem would be of great value for the community, we argue that it falls out of scope of the PySDM JOSS paper for three reasons:
- none of the authors on the paper has enough experience with pyrcel or PyBox,
- PySDM focus, as the name suggests, is on the Monte-Carlo representation of coalescence (condensation/evaporation are of course of equal importance for using it for modelling clouds, yet focusing on parcel frameworks/condensation only would not match project scope),  
- the paper already exceeds JOSS suggested length of 250-1000 words.
We thus suggest to follow up together with the other package developers and carry out such comparison elsewhere.

Last but not least, let me note and ask for confirmation that we have added Oleksii Bulenok @abulenok to the paper author list. Oleksii has worked on the export mechanisms in PySDM and his VTK exporter PR has been merged before the v1.4 release offering the PySDM users to smoothly save and view multi-dimensional simulation state in such tools as Paraview (now mentioned in the text).

**Comment by slayoo** [comment]
For the record, PySDM v1.4 is now available on PyPI as well: https://pypi.org/project/PySDM/

**Comment by dhhagan** [comment]
👋  @andreas-h, @josephhardinee, @ashwinvis : are you available and interested in reviewing this paper for JOSS?

**Comment by josephhardinee** [comment]
I could probably help generically on the python and modeling side, but let me check if a colleague with more domain knowledge, especially around chemistry, would be interested in reviewing for JOSS first.

**Comment by josephhardinee** [comment]
@dhhagan One is overbooked right now and the other has a COI so I can act as a second reviewer for this.

**Comment by dhhagan** [comment]
@josephhardinee Thanks! I will edit the checklist above and invite you.

**Comment by dhhagan** [comment]
@whedon add @josephhardinee as reviewer

**Comment by whedon** [comment]
OK, @josephhardinee is now a reviewer

**Comment by darothen** [comment]
Hi @dhhagan, can you advise on next steps here? I want to keep this on my radar.

**Comment by dhhagan** [comment]
Hey @darothen thanks for checking in! As of now, we're just waiting on the review from @josephhardinee. I believe the authors made changes in response to your initial comments, so you can either review and update your checklist now or wait until the final revision comes in. Everything is a bit slow right now during the pandemic, but thanks for being on top of it!

**Comment by darothen** [comment]
Thanks @dhhagan, I'll wait until the final revision.

**Comment by dhhagan** [comment]
Hi @josephhardinee - just wanted to ping you one more time to see if you still have this on your radar before moving on. Please let me know if this review is not within your scope? Thanks!

**Comment by josephhardinee** [comment]
Sorry, I accepted this then changed jobs and this fell off radar. I will get this done this week.

**Comment by dhhagan** [comment]
No problem, @josephhardinee - thanks so much for the quick reply!

**Comment by slayoo** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by slayoo** [comment]
(API and affiliation updates)

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
(added mention of one additional example based on Lowe et al. 2019 and added Clare Singer, who contributed the example, to the author list)

**Comment by josephhardinee** [comment]
I've completed the review (Sorry it took so long, the package was quite comprehensive). Everything looks good with possibly one exception. The API level documentation is lacking detail once you get below the top level with most methods having no documentation, especially on inputs/outputs (See image below for instance). However, the authors have a fairly robust set of example notebooks that I would argue serves as decent documentation(https://github.com/atmos-cloud-sim-uj/PySDM-examples). I do think we should leave a note encouraging authors to expand API level docs. With that note I would be okay approving this if that meets JOSS standards as the authors appear to be very active in development. @dhhagan would this be acceptable from an editorial viewpoint? 

![image](https://user-images.githubusercontent.com/761704/136965772-661a7e2f-60a3-40a8-9423-cb276db3d7fa.png)

**Comment by darothen** [comment]
@dhhagan Any immediate action items here that you need from the reviewers?

**Comment by slayoo** [comment]
@darothen @dhhagan Let me just confirm, we are actively cleaning up the code towards 100% pylint-clean (incl. docstrings): 
- https://github.com/atmos-cloud-sim-uj/PySDM/pull/663
- https://github.com/atmos-cloud-sim-uj/PySDM/pull/664

Still, we need some days to finish it up.

BTW, we have recently rolled out a proof-of-concept coupler for ClimateMachine.jl and PySDM (with 2D prescribed-flow test case so far featuring PySDM particle-based microphyics and ClimateMachine's fluid flow):
https://github.com/CliMA/ClimateMachine.jl/pull/2244

**Comment by dhhagan** [comment]
Sounds good @slayoo - thanks for the update!

**Comment by slayoo** [comment]
@whedon check references from branch JOSS

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch JOSS
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1002/qj.441 is OK
- 10.1145/2833157.2833162 is OK
- 10.5194/npg-24-535-2017 is OK
- 10.5194/gmd-8-1677-2015 is OK
- 10.5194/acp-18-7313-2018 is OK
- 10.5194/gmd-10-1817-2017 is OK
- 10.21105/joss.00755 is OK
- 10.1029/2018MS001285 is OK
- 10.5194/gmd-13-5119-2020 is OK
- 10.5194/gmd-13-1335-2020 is OK
- 10.1002/2017MS000930 is OK
- 10.5194/gmd-11-3623-2018 is OK
- 10.1002/qj.1913 is OK
- 10.1029/2002JD002673 is OK
- 10.1007/s10546-020-00595-w is OK
- doi:10.1002/fld.1071 is OK
- 10.1029/2019MS001689 is OK
- 10.21105/joss.02807 is OK
- 10.1038/s41467-019-12982-0 is OK

MISSING DOIs

- 10.1007/978-3-030-77964-1_2 may be a valid DOI for title: On the design of Monte-Carlo particle coagulation solver interface: a CPU/GPU Super-Droplet Method case study with PySDM

INVALID DOIs

- 1520-0469(1967)024<0688:CDGBC>2.0.CO;2 is INVALID
```

**Comment by slayoo** [comment]
@whedon check references from branch JOSS

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch JOSS
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1002/qj.441 is OK
- 10.1145/2833157.2833162 is OK
- 10.5194/npg-24-535-2017 is OK
- 10.5194/gmd-8-1677-2015 is OK
- 10.1175/1520-0469(1967)024<0688:CDGBC>2.0.CO;2 is OK
- 10.5194/acp-18-7313-2018 is OK
- 10.5194/gmd-10-1817-2017 is OK
- 10.21105/joss.00755 is OK
- 10.1029/2018MS001285 is OK
- 10.5194/gmd-13-5119-2020 is OK
- 10.5194/gmd-13-1335-2020 is OK
- 10.1002/2017MS000930 is OK
- 10.5194/gmd-11-3623-2018 is OK
- 10.1007/978-3-030-77964-1_2 is OK
- 10.1002/qj.1913 is OK
- 10.1029/2002JD002673 is OK
- 10.1007/s10546-020-00595-w is OK
- doi:10.1002/fld.1071 is OK
- 10.1029/2019MS001689 is OK
- 10.21105/joss.02807 is OK
- 10.1038/s41467-019-12982-0 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by slayoo** [comment]
@whedon generate pdf from branch JOSS

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
> We have already set up a pylint workflow for the project which gives us indications where docstrings (and other things) are missing or need improvement. Will now focus on improving it and report back here within a few weeks.

Dear @dhhagan, @josephhardinee & @darothen,

As of PySDM v1.24 (released just now) 100% of PySDM Python modules have docstrings included. These are top-file module docstrings, mostly 1-2 lines, not yet a full class/method coverage, but given that most of PySDM source files are single-class modules, this should give the users/contributors a meaningful overview of the project when browsing the online docs at https://atmos-cloud-sim-uj.github.io/PySDM.

Moreover, the presence of the module-level docstrings is now ensured at CI level for all PRs using `pylint`, so new code will follow suit. The documentation is generated also during CI builds (and deployed when merging to `master`), with `python -We -m pdoc`, causing any code reference mismatches within the docstrings to be detected and flagged by triggering CI failure. Thus, refactors missing docstring updates will not pass unnoticed.

The documentation sprint has been an occasion for significant cleanups and some refactors aimed at simplifying directory structure and API (all reflected in tests, examples and README Python/Julia/Matlab snippets - as all are checked as a part of CI builds). The code snippets in the paper have also been updated (public API changes were minor, mainly requiring updates of import lines).

We are planning further steady buildup of the docstrings coverage as well as rollouts of other user-oriented resources. 

We hope that the reached interim milestone of 100% module docstring coverage and the introduced CI documentation-related automation can serve as a solid indication of our commitment to maintain and enhance PySDM documentation. Looking forward to your comments and acceptance of the paper for publication in JOSS.

Thank you for your consideration,
Sylwester

**Comment by josephhardinee** [comment]
I'd like to commend the authors for seeing through their commitment to improve documentation as well as the accompanying code cleanups. This looks good to me and I think this greatly deserves publication in JOSS.

**Comment by slayoo** [comment]
thank you, @josephhardinee

**Comment by dhhagan** [comment]
@slayoo Great to see all of the improvements. @darothen how does it look? Are you ready to check off the last few boxes?

**Comment by slayoo** [comment]
For the record, from now on (https://github.com/atmos-cloud-sim-uj/PySDM/pull/717) the code snippets included in the JOSS paper will be executed as a part of PySDM CI build to ensure the example code works well with future updates to PySDM.

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch master. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch master. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3219 with the following error: 
```
 Error reading bibliography file paper.bib:
(line 162, column 3):
unexpected "u"
expecting space, ",", white space or "}"
Looks like we failed to compile the PDF

```

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch master. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch master. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by dhhagan** [comment]
Hey @darothen - just wanted to ping you one more time now that the holidays are over. Is there anything else you need from the authors to finish up this submission?

**Comment by dhhagan** [comment]
@whedon remind @darothen in 1 day

**Comment by whedon** [comment]
Reminder set for @darothen in 1 day

**Comment by whedon** [comment]
:wave: @darothen, please update us on how your review is going (this is an automated reminder).

**Comment by dhhagan** [comment]
Hey @slayoo I am going to try to reach Dan R. offline to see if he can finish this up before needing to find another reviewer. Sorry for the delay here!

**Comment by darothen** [comment]
Sorry for the delay folks; between job change + holidays + AMS it took far longer than I had hoped to check in here.

I reviewed all of the changes and the discourse from the past few months. I think all of my comments/recommendations have either been directly addressed or addressed in the spirit of the substantial improvements and additions made to the library since the first version I reviewed last year. In my opinion, the work here is excellent and is ready for publication in JOSS without any additional modifications.

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch master. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03219/joss.03219/10.21105.joss.03219.pdf) :page_facing_up: :point_left:

**Comment by slayoo** [comment]
@dhhagan hope all is OK. If possible, please update us on the workflow and further requirements given the above messages from the reviewers. Thank you

**Comment by dhhagan** [comment]
Hey @slayoo I should have an update for you soon.

**Comment by dhhagan** [comment]
@slayoo Can you please do the following:

  - [ ] Make a tagged release and archive
  - [ ] Report the version number and archive DOI in this thread

**Comment by slayoo** [comment]
We have moved ahead of the "v1" branch some time ago for this project,
so here are DOIs for Zenodo-archived last release from the v1 branch described in this paper:
- PySDM v1.27: https://doi.org/10.5281/zenodo.6321270
- PySDM-examples v1.27: https://doi.org/10.5281/zenodo.6321279

Thank you

**Comment by dhhagan** [comment]
@slayoo Can you update the title and author list in zenodo to make it match the paper?

**Comment by dhhagan** [comment]
@editorialbot set v1.27 as version

**Comment by slayoo** [comment]
just added one more funding acknowledgement to the paper.md

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by slayoo** [comment]
nice to meet you, @editorialbot

**Comment by slayoo** [comment]
@editorialbot generate pdf from branch master

**Comment by slayoo** [comment]
@editorialbot commands

**Comment by slayoo** [comment]
@editorialbot generate pdf

**Comment by slayoo** [comment]
@dhhagan the markdown-to-latex-to-pdf workflow makes it tricky for us to control line breaks in the list of authors and affiliations:
![image](https://user-images.githubusercontent.com/89685/156810607-9a73177e-846f-4fe6-aba2-8ed67eaafc7e.png)

Some kludge workarounds seem to be possible by inserting non-braking spaces here and there, but it is cumbersome, and I'm not sure if it will actually even be taken into account when producing the final pdf - please confirm if we should "fight" these problems ourselves or if there is any final hand-tuning done for such aspects at JOSS. Thanks!

**Comment by dhhagan** [comment]
Hey @piotrbartman It looks like the title still doesn't match in the archive. Can you please change it to "PySDM v1: particle-based cloud modelling package for warm-rain microphysics and aqueous chemistry".

May also want to remove one of the "l"'s in "modeling".

**Comment by slayoo** [comment]
US spelling introduced here: https://github.com/atmos-cloud-sim-uj/PySDM/commit/dfdfc3a22a051ee7e761bbe4207cc27f2b120301

Zenodo titles updated:
- https://zenodo.org/record/6321270
- https://zenodo.org/record/6321280

Thanks!

**Comment by dhhagan** [comment]
@editorialbot  recommend-accept

**Comment by slayoo** [comment]
@dhhagan, let me ask for your action in resolving the above bot reply. Thank you for your help

**Comment by slayoo** [comment]
@dhhagan, just noticed one more issue: the archive URL posted by the @editorialbot above is a broken link, apparently a DOI was expected as it gives the following URL: "https://doi.org/https://zenodo.org/record/6321270"

**Comment by dhhagan** [comment]
@slayoo Sorry about that -> trying to clear these up now.

**Comment by dhhagan** [comment]
@editorialbot set 10.5281/zenodo.6321270 as archive

**Comment by slayoo** [comment]
@editorialbot commands

**Comment by dhhagan** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@dhhagan – this is failing because the DOI in the BibTeX here is invalid (the preceding `doi:` needs removing): https://github.com/atmos-cloud-sim-uj/PySDM/blob/master/paper/paper.bib#L170

**Comment by slayoo** [comment]
@editorialbot check references

**Comment by claresinger** [comment]
@slayoo it seems the DOI being accessed is not the same as what is in the .bib file https://github.com/atmos-cloud-sim-uj/PySDM/blob/master/paper/paper.bib#L48
`10.1175/1520-0469(1967)024<0688:CDGBC>2.0.CO;2`

**Comment by slayoo** [comment]
@claresinger, thanks, yes, it seems that recently something has changed in the way the bot parses the bib file, I've also just noted it in the PyMPDATA paper: https://github.com/openjournals/joss-reviews/issues/3896#issuecomment-1101700414

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
> ⚠️ [Error prepararing paper acceptance](https://github.com/openjournals/joss-papers/actions/runs/2189150869?check_suite_focus=true). The generated XML metadata file is invalid.

@tarleb – any ideas what this kind of error is about?

**Comment by tarleb** [comment]
The figures are defined using LaTeX syntax. We try hard to parse and convert LaTeX as much as sensible to accommodate cases like this, but that appears one of the cases that we miss.

The solution would be to use pandoc's "implicit figure" feature: a paragraph that contains just an image will be treated as a figure. E.g.

``` markdown
![Sample plot generated with the code snippets included in the paper.](readme.pdf){#fig:readme_fig_1 width="60%"}
```

**Comment by slayoo** [comment]
@whedon generate pdf from branch master

**Comment by slayoo** [comment]
@editorialbot generate pdf from branch master

**Comment by slayoo** [comment]
@editorialbot generate pdf

**Comment by slayoo** [comment]
> This happens as the referenced figure is not included in the JATS output.
> 
> The figures are defined using LaTeX syntax. We try hard to parse and convert LaTeX as much as sensible to accommodate cases like this, but that appears one of the cases that we miss.
> 
> The solution would be to use pandoc's "implicit figure" feature: a paragraph that contains just an image will be treated as a figure. E.g.
> 
> ```gfm
> ![Sample plot generated with the code snippets included in the paper.](readme.pdf){#fig:readme_fig_1 width="60%"}
> ```

changed here: https://github.com/atmos-cloud-sim-uj/PySDM/commit/188609b26f86956602b0c569057a1bbfdd4f6d63
thank you @tarleb & @arfon !

**Comment by slayoo** [comment]
@dhhagan, @arfon, let me ask if you could retry the needed bot command to check if the "Error prepararing paper acceptance" still shows up (BTW, reported the error message typo here: https://github.com/openjournals/joss-papers/pull/3165), thanks.

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@darothen, @josephhardinee – many thanks for your reviews here and to @dhhagan for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@piotrbartman – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by slayoo** [comment]
@arfon Thank you!
In the final paper page that one is redirected to by using the DOI the author list is full of unescaped HTML non-braking space:
![image](https://user-images.githubusercontent.com/89685/164995668-772e2028-dd78-4b12-915f-58f3737dfe90.png)
Is it possible to do correct it?
Thanks

**Comment by arfon** [comment]
@slayoo - please remove all of the &nsbp; characters and I'll re-deposit the paper.

**Comment by piotrbartman** [comment]
@dhhagan @darothen @josephhardinee @arfon @slayoo Thank you all! 🎉🎉🎉

**Comment by slayoo** [comment]
> @slayoo - please remove all of the &nsbp; characters and I'll re-deposit the paper.

Removed: https://github.com/atmos-cloud-sim-uj/PySDM/commit/cf59297dbe6728d479e0b8d2224903a71b37db70

(note, we have changed the branch name from `master` to `main`)

Thank you, @arfon!

**Comment by arfon** [comment]
@editorialbot re-accept

**Comment by arfon** [comment]
@editorialbot reaccept

**Comment by slayoo** [comment]
Thank you, @arfon!

**Comment by arfon** [comment]
:+1: looks like this is now fixed. Closing.
